### PR TITLE
improved isNormal and isGround computations

### PR DESCRIPTION
**Pull request by andreistefanescu**
do not review

### CODE REVIEW

**Comment by andreistefanescu** [comment]
@dwightguth please review

**Comment by dwightguth** [comment]
looks good but there's a couple minor improvements I'd like to see with respect to the isGround and variableSet fields. Namely, if isGround is true, we should be able to deduce that variableSet is empty without traversing the entire term, and similarly, if we have already computed the variableSet, we should be able to determine  isGround without traversing the term by checking the size of the set.

Also, do you have any statistics on the effect this diff has on performance? I'm assuming you did it because it was faster but I'd be interested to see numbers before we merge this.

**Comment by andreistefanescu** [comment]
The old version is about 50% slower. This does not mean much for concrete execution, but it's useful for verification.

File: java-backend/src/main/java/org/kframework/backend/java/kil/JavaSymbolicObject.java [hunk 1]
```diff
@@ -113,39 +110,51 @@ public JavaSymbolicObject substitute(Variable variable, Term term, TermContext c
...
+        return variableSet;
     }
 
-     /**
+    /**
+     * Returns {@code true} if this JavaSymbolicObject does not contain any variables.
+     */
+    public boolean isGround() {
+        if (isGround == null) {
+            if (variableSet == null || !variableSet.isEmpty()) {
```
Comment (dwightguth) [inline java-backend/src/main/java/org/kframework/backend/java/kil/JavaSymbolicObject.java]: `if (variableSet == null) {`

File: java-backend/src/main/java/org/kframework/backend/java/kil/JavaSymbolicObject.java [hunk 2]
```diff
@@ -113,39 +110,51 @@ public JavaSymbolicObject substitute(Variable variable, Term term, TermContext c
...
+                new IsGroundFieldInitializer().visitNode(this);
+            } else {
+                isGround = true;
```
Comment (dwightguth) [inline java-backend/src/main/java/org/kframework/backend/java/kil/JavaSymbolicObject.java]: `isGround = variableSet.isEmpty();`

**Comment by andreistefanescu** [comment]
On a side note, `hashCode` is a lot faster than `isGround`, so all the boilerplate code is slowing things down. Unfortunately, it's not feasible to replace all the visitor/transformer classes with handcrafted code.

**Comment by dwightguth** [comment]
still one minor fix to make, then you can merge.

**Comment by dwightguth** [comment]
lgtm. merge if tests pass.

**Pull request merged**